### PR TITLE
Fixing source map locations when peephole optimizations are applied.

### LINKED ISSUE #14167: [Bug][compiler-v2] Fix source location mapping for peephole optimized code
# 🐛 Bug

After #13894 lands, some of the source location information (file format bytecode code offset to source locations) for peephole optimized code could be broken in the source map.

This can lead to incorrect location information in the error messages seen by the move prover user (see `third_party/move/move-prover/tests/sources/functional/aborts_if.v2_exp`, with the change below).

image

Temporarily, this can be mitigated by turning off peephole optimization (using the env var `MVC_EXP=peephole-optimization=off`) to get correct location information in the errors.

A possible solution to this issue is for the peephole optimizer to also know about the `file_format::CodeOffset -> stackless_bytecode::AttrId` mapping. The individual optimizers can then remapped them as best as possible, which can then reflected back into the source map.

### PR DESCRIPTION
**Pull request by vineethk**
## Description

Fixes #14167.

Peephole optimization changes the original file format bytecode after it is generated from stackless bytecode. Source mapping is computed when the original file format bytecode is created, so the source mapping is incorrect after peephole optimization changes the file format bytecode.

This resulted in several downstream issues, including move test, prover, coverage, etc.

This PR contains the fix by having the peephole optimization maintain a mapping to the original offsets and propagating them, and updating the source map accordingly after the optimization.

## How Has This Been Tested?
- Existing tests
- Added another test reported by a user
- Manually tested the user's report before and after the fix (details below).

When we ran `aptos move test --move-2` for the following program:

```
module anto::Example {
    use std::option::{Option};

    const E_CONDITION_A: u64 = 1;
    const E_CONDITION_B: u64 = 2;
    const E_CONDITION_C: u64 = 3;
    const E_CONDITION_D: u64 = 4;

    fun validate_parameters(
        param_a: Option<u64>, // Optional first parameter to validate
        param_b: Option<u64>, // Optional second parameter to validate
        reference_value: u64, // Reference value for comparisons
        threshold_value: u64, // Threshold for additional validation
        flag: bool // Indicates a positive or negative condition
    ) {
        // Determine the context based on the flag
        let is_positive = flag;

        // Check and validate param_a if it exists
        if (param_a.is_some()) {
            // Extract the value from the Option
            let value_a = param_a.extract();

            // Ensure the value is non-zero
            assert!(value_a > 0, E_CONDITION_A);

            // Validate based on the condition (is_positive)
            let is_valid_a =
                if (is_positive) {
                    value_a > reference_value // For positive condition, value_a must be greater than reference_value
                } else {
                    value_a < reference_value // For negative condition, value_a must be less than reference_value
                };

            // Assert that the validation passed
            assert!(
                is_valid_a,
                if (is_positive) E_CONDITION_B else E_CONDITION_C
            );
        };

        // Check and validate param_b if it exists
        if (param_b.is_some()) {
            // Extract the value from the Option
            let value_b = param_b.extract();

            // Ensure the value is non-zero
            assert!(value_b > 0, E_CONDITION_A);

            // Validate based on the condition (is_positive)
            let is_valid_b =
                if (is_positive) {
                    value_b < reference_value // For positive condition, value_b must be less than reference_value
                } else {
                    value_b > reference_value // For negative condition, value_b must be greater than reference_value
                };

            // Assert that the validation passed
            assert!(
                is_valid_b,
                if (is_positive) E_CONDITION_C else E_CONDITION_D
            );

            // Additional validation against the threshold value if it exists
            if (threshold_value > 0) {
                let is_valid_threshold =
                    if (is_positive) {
                        value_b > threshold_value // For positive condition, value_b must be greater than threshold_value
                    } else {
                        value_b < threshold_value // For negative condition, value_b must be less than threshold_value
                    };

                // Assert that the threshold validation passed
                assert!(is_valid_threshold, E_CONDITION_A);
            }
        };
    }

    #[test_only]
    use std::option;

    #[test]
    fun test_validate_parameters() {
        // Passing Invalid param_a for positive condition
        // This should throw E_CONDITION_B error
        validate_parameters(
            option::some(40), // Less than reference_value (60)
            option::some(50),
            60,
            30,
            true
        );
    }
}
```

on the `main` branch, we get:

```
│  9 │     fun validate_parameters(
│    │         ------------------- In this function in 0xffeeede::Example
│    ·
│ 48 │             assert!(value_b > 0, E_CONDITION_A);
│    │             ^^^^^^ Test was not expected to error, but it aborted with code 2 originating in the module 000000000000000000000000000000000000000000000000000000000ffeeede::Example rooted here
```

with this PR, we get (correctly):

```
│  9 │     fun validate_parameters(
│    │         ------------------- In this function in 0xffeeede::Example
│    ·
│ 36 │             assert!(
│    │             ^^^^^^ Test was not expected to error, but it aborted with code 2 originating in the module 000000000000000000000000000000000000000000000000000000000ffeeede::Example rooted here
```

## Key Areas to Review
- Correctness of the offset propagation
- It is expected that there are no changes to the generated file format bytecode itself, if you find anything that suggests otherwise, please comment.

## Type of Change
- [x] Bug fix

## Which Components or Systems Does This Change Impact?
- [x] Move Compiler

### CODE REVIEW

**Comment by vineethk** [comment]
* **#15384** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15384?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15384?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by rahxephon89 — APPROVED** [review]
LGTM, thanks

File: third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/inefficient_loads.rs [hunk 1]
```diff
@@ -61,8 +64,12 @@ impl WindowOptimizer for InefficientLoads {
                     // We have reached the end of the pattern (point 4 in the module documentation).
                     let sequence = &window[2..index + 2];
                     let load_constant = &window[0..1];
+                    let transformed_code = [sequence, load_constant].concat();
+                    let original_offsets = (2..(index + 2) as CodeOffset)
+                        .chain(iter::once(0))
+                        .collect::<Vec<_>>();
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/inefficient_loads.rs:70]: Could we write this more intelligibly as 
```
let original_offsets = [[std::ops::Range<CodeOffset> { start: 2, end: index+2}, [0]].concat();
```
or perhaps just add a comment:
```
// original_offsets = [2, ..., index+2, 0];
```

**Review by brmataptos — APPROVED** [review]
Looks ok, though I'd at least add a comment for code clarity.

File: third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/inefficient_loads.rs (hunk 1 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/inefficient_loads.rs:72]: done, added a comment

File: third_party/move/move-ir-compiler/move-bytecode-source-map/src/source_map.rs [hunk 2]
```diff
@@ -202,6 +202,29 @@ impl FunctionSourceMap {
...
+            if prev_loc == loc {
+                // optimization: if a series of instructions map to the same location, we only need
+                // to add code mapping for the first one.
+                continue;
+            }
+            if let Some(loc) = loc {
+                self.add_code_mapping(new_offset as CodeOffset, loc);
+                prev_loc = Some(loc);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (vineethk) [reply third_party/move/move-ir-compiler/move-bytecode-source-map/src/source_map.rs]: Took the spirit of the posted code (thanks), and rewrote it a bit.

The code posted has a couple of bugs (eg., when using `scan`, when the closure returns `None`, the iteration ends, which is not what we want here).

**Pull request merged**